source/common/stats/symbol_table_impl.h

```diff
@@ -130,22 +70,94 @@ class SymbolEncoding {
  */
 class SymbolTableImpl : public SymbolTable {
 public:
+  /**
+   * Intermediate representation for a stat-name. This helps store multiple names
+   * in a single packed allocation. First we encode each desired name, then sum
+   * their sizes for the single packed allocation. This is used to store
+   * MetricImpl's tags and tagExtractedName. Like StatName, we don't want to pay
+   * a vptr overhead per object, and the representation is shared between the
+   * SymbolTable implementations, so this is just a pre-declare.
+   */
+  class Encoding {
+  public:
+    /**
+     * Before destructing SymbolEncoding, you must call moveToStorage. This
+     * transfers ownership, and in particular, the responsibility to call
+     * SymbolTable::clear() on all referenced symbols. If we ever wanted
+     * to be able to destruct a SymbolEncoding without transferring it
+     * we could add a clear(SymbolTable&) method.
+     */
+    ~Encoding();
+
+    /**
+     * Encodes a token into the vec.
+     *
+     * @param symbol the symbol to encode.
+     */
+    void addSymbol(Symbol symbol);
+
+    /**
+     * Encodes an entire string into the vec, on behalf of FakeSymbolTableImpl.
+     * TODO(jmarantz): delete this method when FakeSymbolTableImpl is deleted.
+     *
+     * @param str The string to encode.
+     */
+    void addStringForFakeSymbolTable(absl::string_view str);
+
+    /**
+     * Decodes a uint8_t array into a SymbolVec.
+     */
+    static SymbolVec decodeSymbols(const SymbolTable::Storage array, uint64_t size);
+
+    /**
+     * Returns the number of bytes required to represent StatName as a uint8_t
+     * array, including the encoded size.
+     */
+    uint64_t bytesRequired() const { return size() + StatNameSizeEncodingBytes; }
+
+    /**
+     * Returns the number of uint8_t entries we collected while adding symbols.
+     */
+    uint64_t size() const { return vec_.size(); }
+
+    /**
+     * Moves the contents of the vector into an allocated array. The array
+     * must have been allocated with bytesRequired() bytes.
```

Comment:
What happens if the array is undersized?
(Sorry for the comments here, I know you're just moving this around instead of writing it anew.)